### PR TITLE
hare activeset: employ ATX grading scheme

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation
part of #4089

## Changes
- grade atxs as described in https://community.spacemesh.io/t/grading-atxs-for-the-active-set/335, let s be the start of the epoch, and δ the network propagation time.
  - grade 0/evil: ATX was received at time t >= s-3δ, or an equivocation proof was received by time s-δ.
  - grade 1/acceptable: ATX was received at time t < s-3δ before the start of the epoch, and no equivocation proof was received by time s-δ.
  - grade 2/good: ATX was received at time t < s-4δ, and no equivocation proof was received for that id until time s.
 
- miner only include grade 2/good ATXs in its active set for its first ballot of the epoch
- hare only vote for proposals with grade 1/acceptable and 2/good atxs in the activeset
- newly sync'ed node uses the first block of epoch for active set in ref ballot, since it doesn't have accurate received timestamp for atxs or malfeasance proof

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

**Comment by dshulyak** [comment]
so synced node votes against everyone? or it downloads timestamp from a peer?

**Comment by countvonzero** [comment]
> so synced node votes against everyone? or it downloads timestamp from a peer?

timestamps are not synced. so yes. a newly synced node will vote against everyone and end up with an empty active set.

this means
1. a new node (synced from genesis) won't be able to make a block proposal in the same epoch that it join (empty active set)
2. a node that are mostly offline before `s-4δ` won't be able to make a block proposal in the following epoch (empty active set)
3. a node that publishes its atx within `s-4δ` window, won't be able to make a block proposal in the following epoch (the node itself is not in the active set)

in 1 and 2, it also won't vote for any proposals in the hare preround, but this should not affect hare consensus

in essence, this incentivize miners to stay online as much as possible.

**Comment by dshulyak** [comment]
i understand that this is an attempt to solve some problems, but it almost guaranteed to introduce new problems. i don't see this complexity as a good protocol extension, maybe we can find a reference in working implementations that are doing something similar. otherwise i hope someone else can approve it

**Comment by countvonzero** [comment]
> i understand that this is an attempt to solve some problems, but it almost guaranteed to introduce new problems. i don't see this complexity as a good protocol extension, maybe we can find a reference in working implementations that are doing something similar. otherwise i hope someone else can approve it

it's not about approving this PR.... it's more about flushing out concerns you may have.

> maybe we can find a reference in working implementations that are doing something similar

by working implementations do you mean other blockchain projects?

**Comment by dshulyak** [comment]
yes, blockchain or other software (not necessarily byzantine) that works. i reviewed quite a few projects in the last years, and i don't recall conditions that assume so much.

**Comment by dshulyak** [comment]
what do you think about next criteria for making such changes? there must be a test that clearly shows sequence of steps that must result in consensus failure. if the change is so obscure (or misunderstood) that it can't be tested - then there is no reason to make the change. 

i tried to understand Tal explanation, but it literally feels like a random change to me, that may fix something or may not fix anything

File: hare/hare.go [hunk 1]
```diff
@@ -451,12 +460,32 @@ func goodProposals(logger log.Log, msh mesh, nodeID types.NodeID, lid types.Laye
 		ownTickHeight = ownHdr.TickHeight()
 	}
 	for _, p := range props {
+		hdr, err := msh.GetAtxHeader(p.AtxID)
```
Comment (countvonzero) [inline hare/hare.go]: should look at p's active set, not p's own atxid

**Comment by countvonzero** [comment]
merged separately in
https://github.com/spacemeshos/go-spacemesh/pull/4717
https://github.com/spacemeshos/go-spacemesh/pull/4718

**Pull request closed without merging**